contributors/design-proposals/raw-block-pv.md

```diff
@@ -0,0 +1,494 @@
+    pdName: "gce-disk-1"
+```
+
+***If admin specifies volumeType: block + fstype: ext4 then they would have the default behavior of files on block ***
+***fsType values will be provisioner dependent. Block is suggested for development simplicity. Since the PVC object is passed
+   to the provisioner, it will be responsible for validating and handling whether or not it supports the volumeType being passed ***
+
+# Container Runtime considerations
+It is important the values that are passed to the container runtimes are valid and support the current implementation of these various runtimes. Listed below are a table of various runtime and the mapping of their values to what is passed from the mount. 
+
+| runtime engine    | runtime options  | accessMode       |  
+| --------------    |:----------------:| ----------------:|
+| docker/runc/rkt   |  mknod / RWM     | RWO              |
+| docker/runc/rkt   |       R          | ROX              |
+
+The accessModes would be passed as part of the options array and would need validate against the specific runtime engine. 
+Since rkt doesn't use the CRI, the config values would need to be passed in the legacy method.
+
+The runtime option would be placed in the DeviceInfo as such:
+devices = append(devices, kubecontainer.DeviceInfo{PathOnHost: path, PathInContainer: path, Permissions: "mrw"}) for RWO
+
+The implemenation plan would be to rename the current makeDevices ot makeGPUDevices and create a seperate function to add the raw block devices to the option array to be passed to the container runtime. This would interate on the paths passed in for the pod/container.
+
+# Implementation Plan, Features & Milesones
+
+Phase 1: v1.8
+Feature: Pre-provisioned PVs to precreated devices 
+
+               Milestone 1: API changes
+
+               Milestone 2: Restricted Access 
+               
+               Milestone 3: Scrubbing device after PV is deleted
+               
+               Milestone 4: Changes to the mounter interface as today it is assumed 'file' as the default.
+               
+               Milestone 5: Expose volumeType to users via kubectl
+               
+               Milestone 6: Adds enable/disable configuration to securityContext in PSP (Pod Security Policy) similar to hostPath
+               
+               Milestone 7: Validate container runtime options with user specifcations as indicated in UC3
+	       
+	       Milestone 8: Container Runtime changes
+
+Phase 2:  v1.9
+Feature: Discovery of block devices 
+
+                Milestone 1: Dynamically provisioned PVs to dynamically allocated devices
+
+                Milestone 2: Privileged container concerns    
+
+Other considerations:
+
+               Reference volume driver change(Attach/Detach logic) for pre-provisioned PVs
+
+               Reference volume driver change(Provision/Delete logic) for dynamic provisioning
+
+## Mounter interface proposed design
+
+```
+ type BlockMounter interface {
+	Mounter
```

Comment:
@jsafrane 

>> SetUpDevice / TearDownDevice

+1.
I'm not clear whether we need to provide both SetUpDevice(), SetUpDeviceAt().
Are there any reason that the SetUp() and SetUpAt() are separated?

>> Taking iSCSI as an example, the plugin attaches a device /dev/sdb and some links appear in /dev/disk/by-*.

If the volumeType is block, we only call SetUpDevice at GenerateMountVolumeFunc() in operation_generator, therefore iSCSI plugin still needs device attach logic(/dev/sdb) in the SetUpDevice() like this.

```
- SetUpDevice()
  -> SetUpDeviceAt()
     -> diskSetUp() : prepare symlink instead of mount point or bind mount.
         -> AttachDisk() : attach disk to the kubelet
```
>>we should mimic this behavior and put symlinks(?) to the devices into
>> /var/lib/kubelet/pods/<pod uid>/devices/<plugin name>/<volume name> + adjust
>> kubelet's cleanup functions accordingly.

Yes!
Without filesystem, we can't mount the volume to Pod directory and Global mount directory, therefore, I thought we need alternative way using "symlink" as you mentioned.

> `</var/lib/kubelet/pods/<pod uid>/devices/<plugin name>/<volume name>>` + adjust
> kubelet's cleanup functions accordingly.

+1.
I suppose it is easy to find block volumes attached to a container by separating directory from filesystem volume. And I think the <volume name> is a symlink to the "/dev/sdb".